arch/Etca/EtcaDisassembler.c

```diff
@@ -0,0 +1,621 @@
+		ctx->rel.extended = d;
+
+		code += 2;
+		code_len -= 2;
+		(*size) += 2;
+	} else if (code_len && code[0] >> 4 == 0xF &&
+		   code_len >= 1 + (1 << (code[0] & 3))) {
+		ctx->ss = code[0] & 3;
+		int sz = 1 << ctx->ss;
+
+		uint64_t d = 0;
+		for (int i = 0; i < sz; i++) {
+			d <<= 8;
+			d |= code[i + 1];
+		}
+		ctx->rel.present = true;
+		ctx->rel.extended = d;
+
+		// clang-format off
+		switch ((code[0] >> 2) & 3) {
+		case 0: ctx->insn = ETCA_INS_REL_JMP; break;
+		case 1: ctx->insn = ETCA_INS_ABS_JMP; break;
+		case 2: ctx->insn = ETCA_INS_REL_CALL; break;
+		case 3: ctx->insn = ETCA_INS_ABS_CALL; break;
+		}
+		// clang-format on
+
+		if (ctx->insn == ETCA_INS_REL_JMP ||
+		    ctx->insn == ETCA_INS_REL_CALL)
+			sign_extend(d, uint64_t, sz * 8);
+
+		code += sz + 1;
+		code_len -= sz + 1;
+		(*size) += sz + 1;
+	} else if (code_len >= 2 && code[0] >> 7 == 0 &&
+		   code[0] >> 2 != 7 /* 0b111 */) {
+		bool imm = code[0] & (1 << 6);
+
+		if (imm) {
+			parseRI(ctx, code[1], ctx->insn);
+		} else {
+			parseABM(ctx, code[1]);
+		}
+
+		/* abm with fi is treated as imm */
+		if (!imm && ctx->abm.m == 1 &&
+		    (ctx->abm.b == 2 || ctx->abm.b == 3)) {
+			imm = true;
+		}
+
+		ctx->insn = parseBaseOpcode(code[0] & 0xF, imm);
+		if (ctx->insn == ETCA_INS_INVALID)
+			return false;
+
+		ctx->ss = (code[0] >> 4) & 3;
+
+		code += 2;
+		code_len -= 2;
+		(*size) += 2;
+	} else {
+		return false;
+	}
+
+	*code_len_p = code_len;
+	*code_p = code;
+
+	return true;
+}
+
+// returns true if valid
+bool Etca_getInstruction(csh ud, const uint8_t *code, size_t code_len,
+			 MCInst *mcInstr, uint16_t * /* out */ size,
+			 uint64_t address, void *infoIn)
+{
+	etca_info *info = infoIn;
+	// cs_detail *detail = mcInstr->flat_insn->detail;
+
+	DecodeIsntCtx ctx = { 0 };
+	ctx.insn = ETCA_INS_INVALID;
+	ctx.cond = ETCA_COND_ALWAYS;
+	*size = 0;
+
+	/* conditional prefix */
+	if (code_len && code[0] >> 4 == 2 + 8 /* 0b1010 */ &&
+	    (code[0] & 0xF) < 14) {
+		// cond prefix of always / never isn't allowed
+		ctx.pfx_cond.present = true;
+		ctx.cond = code[0] & 0xF;
+		code++;
+		code_len--;
+		(*size)++;
+	}
+
+	/* register expansion prefix */
+	if (code_len && code[0] >> 4 == 4 + 8 /* 0b1100 */) {
+		ctx.pfx_rex.present = true;
+		ctx.pfx_rex.q = (code[0] >> 3) & 1;
+		ctx.pfx_rex.a = (code[0] >> 2) & 1;
+		ctx.pfx_rex.b = (code[0] >> 1) & 1;
+		ctx.pfx_rex.x = (code[0] >> 0) & 1;
+		code++;
+		code_len--;
+		(*size)++;
+	}
+
+	// if (code_len && code[0] >> 4 == 1 + 4 + 8 /* 0b1101 */)
+	//    unused prefix
+
+	if (!parseCoreOp(&ctx, &code, &code_len, size))
+		return false;
+
+	if (ctx.abm.present && ctx.abm.m == 1 && ctx.abm.b == 2) {
+		/* 1B full immediate */
+		if (!(code_len >= 1))
+			return false;
+		ctx.abm.present = false;
+		ctx.ri.present = true;
+		ctx.ri.r = ctx.abm.a;
+		ctx.ri.imm = code[0];
+
+		code += 1;
+		code_len -= 1;
+		(*size) += 1;
+	} else if (ctx.abm.present && ctx.abm.m == 1 && ctx.abm.b == 3) {
```

Comment:
Please move the magic numbers here to an enum so future devs can easier grasp what they mean.